compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -100,12 +100,34 @@ object Applications {
       Nil
     }
 
+    def unapplySeqTypeElemTp(getTp: Type): Type = {
+      val lengthTp = ExprType(defn.IntType)
+      val lengthCompareTp = MethodType(List("len".toTermName))(_ => defn.IntType :: Nil, _ => defn.IntType)
+      def applyTp(elemTp: Type) = MethodType(List("i".toTermName))(_ => defn.IntType :: Nil, _ => elemTp)
+      def dropTp(elemTp: Type) = MethodType(List("n".toTermName))(_ => defn.IntType :: Nil, _ => defn.SeqType.appliedTo(elemTp))
+      def toSeqTp(elemTp: Type) = ExprType(defn.SeqType.appliedTo(elemTp))
+
+      val elemTp = getTp.member(nme.apply).suchThat(_.info <:< applyTp(WildcardType)).info.resultType
+
+      def names1 = List(nme.lengthCompare, nme.apply, nme.drop, nme.toSeq)
+      def types1 = List(lengthCompareTp, applyTp(elemTp), dropTp(elemTp), toSeqTp(elemTp))
+
+      def names2 = List(nme.length, nme.apply, nme.drop, nme.toSeq)
+      def types2 = List(lengthTp, applyTp(elemTp), dropTp(elemTp), toSeqTp(elemTp))
+
+      val valid = elemTp.exists &&
+                  (getTp <:< RefinedType.make(defn.AnyType, names1, types1) ||
+                  getTp <:< RefinedType.make(defn.AnyType, names2, types2))
+
+      if (valid) elemTp else NoType
+    }
+
+    def validUnapplySeqType(getTp: Type): Boolean = unapplySeqTypeElemTp(getTp).exists
+
     if (unapplyName == nme.unapplySeq) {
-      if (unapplyResult derivesFrom defn.SeqClass) seqSelector :: Nil
-      else if (isGetMatch(unapplyResult, pos) && getTp.derivesFrom(defn.SeqClass)) {
-        val seqArg = getTp.elemType.hiBound
-        if (seqArg.exists) args.map(Function.const(seqArg))
-        else fail
+      if (isGetMatch(unapplyResult, pos) && validUnapplySeqType(getTp)) {
+        val elemTp = unapplySeqTypeElemTp(getTp)
```

Comment:
`unapplySeqTypeElemTp` is also called by `validUnapplySeqType` and seems very expensive. There is no need to call it twice. We could rewrite to something like:
```scala
if (isGetMatch(unapplyResult, pos)) {
  val elemTp = unapplySeqTypeElemTp(getTp)
  if (elemTp.exists) ...
  else fail
}
else fail
```